### PR TITLE
Add #pragma pinvoke

### PR DESCRIPTION
**Pull request by BadRyuner**
Yes, this is not a good implementation of "#pragma pinvoke"... (#511)
Some methods I didn't name correctly, hope for some help with that!

But it's working just fine.

### CODE REVIEW

**Comment by BadRyuner** [comment]
Need a good example of a pinvoke function call that is on all three platforms in the same library... On Windows "puts" function in mscvrt, on Linux "puts" function in libc...

**Comment by BadRyuner** [comment]
Well, the tests passed.

File: Cesium.CodeGen/Contexts/GlobalConstructorScope.cs [hunk 1]
```diff
@@ -77,4 +80,30 @@ public Instruction ResolveLabel(string label)
...
+
+    public T? GetSpecialEffect<T>()
+    {
+        for(int i = _specialEffects.Count - 1; i >= 0; i--)
+        {
+            var effect = _specialEffects[i];
+            if (effect is T t)
+                return t;
+        }
+        return default;
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Contexts/GlobalConstructorScope.cs:94]: Would this work?
```suggestion
        return _specialEffects.OfType<T>().FirstOrDefault();
```

File: Cesium.CodeGen/Contexts/IDeclarationScope.cs [hunk 2]
```diff
@@ -45,6 +46,10 @@ internal interface IDeclarationScope
     /// </summary>
     /// <returns>Name of the virtual label which can be used by continue statement</returns>
     string? GetContinueLabel();
+    
+    void PushSpecialEffect(object declaration);
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Contexts/IDeclarationScope.cs:50]: Please add a comment indicating our pragma as an example of this "special effect", for future readers.

File: Cesium.CodeGen/Contexts/Meta/FunctionInfo.cs [hunk 3]
```diff
@@ -19,6 +20,8 @@ internal record FunctionInfo(
     public StorageClass StorageClass { get; set; } = StorageClass;
     public bool IsDefined { get; set; } = IsDefined;
     public string? CliImportMember { get; set; }
+    public string? DllLibraryName { get; set; }
+    public string? DllImportNameStrip { get; set; }
```
Comment (ForNeVeR) [inline Cesium.CodeGen/Contexts/Meta/FunctionInfo.cs:24]: Let's add documentation referencing `PInvokeDefinition.Prefix`, for the reader to get why are there two properties.

File: Cesium.Preprocessor/CPreprocessor.cs [hunk 4]
```diff
@@ -284,6 +298,12 @@ await foreach (var item in subProcessor.GetPreprocessingResults())
         yield return new Token<CPreprocessorTokenType>(new Range(), new Location(), "\n", NewLine);
     }
 
+    private static IEnumerable<IToken<CPreprocessorTokenType>> TokenizeString(string code)
+    {
+        var tokenizer = new CPreprocessorLexer("c:/null.c", code);
```
Comment (ForNeVeR) [inline Cesium.Preprocessor/CPreprocessor.cs:303]: Let's either mark this as a `null` path (making the parameter nullable), or use some more specific name, like `<null>` or something.

**Review by ForNeVeR — CHANGES_REQUESTED** [review]
Great work! I've left some comments and ready to discuss minor issues.

File: Cesium.CodeGen.Tests/CodeGenPinvokeTests.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            var preprocessor = new CPreprocessor(
+                _mainMockedFilePath,
+                lexer,
+                includeContext,
+                definesContext,
+                warningProcessor);
+            var result = await preprocessor.ProcessSource();
+            return result;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (BadRyuner) [reply Cesium.CodeGen.Tests/CodeGenPinvokeTests.cs:56]: This is from another test: [Cesium.Parsers.Tests->PreprocessorTests](https://github.com/ForNeVeR/Cesium/blob/63bd762e43f52dfd6cfb4ed1478fe190f84cd1fc/Cesium.Parser.Tests/PreprocessorTests/PreprocessorTests.cs#L26)
They don't overlap

**Review by ForNeVeR — APPROVED** [review]
Thanks!

**Pull request merged**